### ISSUE TITLE
`hist` table: useful for automatic replacement of `r` by `r.rotunda`, `s` by `long_s`, etc?

### ISSUE THREAD

**Issue by ctrlcctrlv** [ctrlcctrlv/kjv1611#4]
I was sent a feature request in e-mail by a Mr. Kessler, the automatic replacement using OpenType alternates of `r` by `r.rotunda` and `s` by `long_s`, and certainly all of the ligatures would also apply. `s h` to `long_s h` to `long_s_h` where appropriate.

One difficulty is that both of these switches are not universally applied, but rather applied only in certain circumstances.

I think this might be doable inside the `hist` feature tag, but am not sure. Further research needed.

**Comment by khaledhosny** [comment]
I think the rules for these vary over time and place, so there is no universal way to it which I think why Unicode has dedicated characters for long s and r rotunda, as doing this at text level is the preferred solution.

**Comment by ctrlcctrlv** [comment]
I'd say that for this font any historical OpenType features would be assumed to be following the rules of the book it came from, 1611 King James Bible, so that's a non-issue.

To typeset with other rules, disable the feature and use the Unicode points. No?

**Comment by khaledhosny** [comment]
If the intention is to faithfully represent 1611 King James Bible, then I guess yes, but if it is meant to be a general purpose fraktur font then the answer is probably no. But since `hist` is not on by default, it is probably OK either way as one will have to explicitly enable it and thus should be prepared to handle the consequences.

**Comment by artistofmind** [comment]
I thought of one more thing this feature should do: automatically replace all capital "I"s with capital "J"s.

I actually haven't yet found an "I" in the 1611 KJB. (Save of course in those portions rendered in Garamond, including wherever "J"s would ordinarily go!) I believe there was only one character for both letters in either font. (The lowercase "j" is a final "i," viz. iij for numeral 3.) In fact, at the time, I'm pretty sure there was only one letter in English for both roles. The "J" in the Garamond happens to look like what we consider an "I" now, and the "I" in the blackletter happens to look like what we've since decided (arbitrarily) is a "J."

This is why the Latin form of Yeshua, Iesus ("yay'zoos"), later became mispronounced Jesus ("djee'zuss"); Yakob → Jacob; YHWH → Jehovah; etc. Good English schoolchildren learned that "J" is pronounced differently than "I," some centuries after the convention of writing those names that way was well established. (Meanwhile, in Spanish, they pronounce "j" like an "h." . . .)

Anyway, so I'm genuinely curious: did you invent the "I" in your font, or is it actually used somewhere?

**Comment by ctrlcctrlv** [comment]
It is not actually used. Per [README.md](https://github.com/ctrlcctrlv/kjv1611#the-below-text-is-for-screen-readers-and-search-engines), "V", "I" and "@" are inventions. Many other glyphs are also inventions, perhaps I should make a full list.

**Comment by artistofmind** [comment]
In that case, when the historical / discretionary option is enabled, "V" should also be replaced by "U." :)

**Comment by ctrlcctrlv** [comment]
Good idea. These are the rules I've come up with so far:
```
# TODO: Replace most of these with classes: @a; @c, etc.
@lower_no_s = [a b c d e f g h i j k l m n o p q r t u v w x y z];
#@B
@D = [D Eth Dcaron Dcroat D.alt];
@P = [P Thorn];
#@b
@d = [d dcroat dcaron eth];
@h = [h hcircumflex hbar];
@o = [o ograve oacute ocircumflex otilde odieresis oslash omacron obreve ohungarumlaut oslashacute];
@p = [p thorn];
@w = [w wcircumflex wgrave wacute];
@y = [y yacute ydieresis ycircumflex ygrave];

feature hist {
    lookup simplecases {
        sub V' by U;
        sub I' by J;
    } simplecases;

    lookup longsfix1 {
        sub s' @lower_no_s by longs;
        sub s' s by longs;
    } longsfix1;
    
    lookup longsfix2 {
        # sub longs longs' @lower_no_s by s;
        sub longs' [f f_f] by s;
        sub [f f_f] longs' by s;
    } longsfix2;
    
    lookup rrotundafix1 {
        sub [B @D @P b @d @h @o @p @w @y] r' by r.rotunda;
    } rrotundafix1;
} hist;
```

**Comment by ctrlcctrlv** [comment]
Do you happen to know if two r rotundas in sequence result in an r rotunda?

Also, I *did not actually look at the Bible to compose these rules*, I just used some excellent articles by Andrew West on the subject. ([1](http://babelstone.blogspot.com/2006/07/r-rotunda-part-1.html), [2](http://babelstone.blogspot.com/2006/06/rules-for-long-s.html))

**Comment by artistofmind** [comment]
In the KJB, when there are two "r"s and the first is an r rotunda (such as in "sorry" or "porridge"), the second "r" is a regular "r." Only the first is an r rotunda. It's essentially a calligraphical ligature, which always follows rounded right-facing strokes (as opposed to straight lines or gaps, or anything else really). So while it may look funny, you will sometimes see both in one word, used back to back.

That list looks correct. If I notice any additional precursor letters, I'll be sure to inform you here. :)

**Comment by Alice196498** [comment]
For the long s, I [have written about the rules](https://imgur.com/gallery/0sVAa) in Roman type, but the rules may be a little different in Blackletter ; if it would be helpful, I ſhould gladly compile a liſt of the rules for you—I can do the ſame for the R Rotunda, but it ſeems that the work for that has perhaps been done well enough by others.
 As for V and U : theſe were conſidered to be one letter repreſented by different ſhapes in different poſitions, ſimilar to the difference between ſ and s. In the capital, there was a letter ſimilar in form to our U, which ſtood for both U and V—in the minuſcule, there was a u, which was uſed in medial and final poſitions for both v and u, and a v, which would be uſed in word-initial poſitions ; as, very, vnto, euery, heauen. About I and J : Theſe two were conſidered to be the ſame letter as well, but only was J ever uſed as capital ; there was no capital I at the time—there was alſo no minuſcule j uſed where we might expect it, as it was conſidered to be a variant much like v in the u/v diſtinction above, which was uſed when there were two i’s in a word or in Roman numerals ; as, perij, xiij.
 I believe it would be the beſt practice to ſet theſe as Styliſtick Sets, but they may alſo be bundled together in the hiſtorical feature ſet. I could come up with the needful code if neceſſary, but that may be better left to ſomeone more ſkilled at that than I.

**Comment by ctrlcctrlv** [comment]
Implemented in the latest OTF release (3.1)
https://github.com/ctrlcctrlv/kjv1611/blob/master/features.fea
